rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,334 @@
+
+The supported entrypoints will be:
+- `generate_keypair`
+- `validate_keypair --keypair={keypair}`
+- `gossip_network`
+
+#### Protocol
+
+When the lip2p helper process is first started by the Daemon (in `gossip_network` mode), the daemon will send an `init(config Config)` is written once over stdin. The information sent in this message could theoretically be passed via the CLI arguments, but doing this would lose some type safety, so we prefer to send this data over as an IPC message. Once this message has been received by the helper process, the helper process will open ports, join the network, and begin participating in the main protocol loop. In this main protocol loop, either process should expect to receive any IPC messages over any data streams at any time.
+
+Here is a list of the IPC messages that will be supported in each direction, along with some relevant type definitions in Go:
+
+```txt
+Daemon -> Helper
+  // it's possible to just remove this message as it is just a specialized case of `sendRequests`
+  sendRequestToPeer(requestId RequestId, to Peer, msgType MsgType, rawData []byte)
+  sendRequests(requestId RequestId, to AbstractPeerGraph, msgType MsgType, rawData []byte)
+  sendResponse(requestId RequestId, status ValidationStatus, rawData []byte)
+  broadcast(msgType MsgType, rawData []byte)
+  validate(validation ValidationHandle, status ValidationStatus)
+
+Helper -> Daemon
+  handleRequest(requestId RequestId, from Peer, rawData []byte)
+  handleResponse(requestId RequestId, validation ValidationHandle, rawData []byte)
+  handleGossip(from Peer, validation ValidationHandle, rawData []byte)
+  stats(stats Stats)
+```
+
+```go
+// == The `Config` struct is sent with the `init` message at the start of the protocol.
+
+// The following old fields have been completely removed:
+//   - `metricsPort` (moving over to push-based stats syncing, where we will sync any metrics we want to expose)
+//   - `unsafeNotTrustIp` (only used as a hack in old integration test framework; having it makes p2p code harder to reason about)
+//   - `gaterConfig` (moving towards more abstracted interface in which Go manages gating state data)
+type Config struct {
+  networkId           string      // unique network identifier
+  privateKey          string      // libp2p id private key
+  stateDirectory      string      // directory to store state in (peerstore and dht will be stored/loaded from here)
```

Comment:
it might be nice to store the key in a file in the `stateDirectory`, then the libp2p helper can just check that directory for the file

Reply:
I believe we allow users to pass in their own pre-generate keypair via the CLI. We could still pull from the filesystem, but I figured managing that extra state wasn't worth it if we would just write that keypair file before we spawn the helper each time.